noir_stdlib/src/uint128.nr

```diff
@@ -91,23 +99,36 @@ impl U128 {
         if ascii < 58 {
             ascii - 48
         } else if ascii < 71 {
+            assert(ascii >= 65); // enforce >= 'A'
             ascii - 55
         } else {
+            assert(ascii >= 97); // enforce >= 'a'
+            assert(ascii <= 102); // enforce <= 'f'
             ascii - 87
         } as Field
     }
 
+    // TODO: Replace with a faster version. 
+    // A circuit that uses this function a lot will be really slow to compute
+    // (we're doing binary search to compute the quotient)
     unconstrained fn unconstrained_div(self: Self, b: U128) -> (U128, U128) {
-        if self < b {
-            (U128::from_u64s_le(0, 0), self)
+        if b == U128::zero() {
+            // Return 0,0 to avoid eternal loop
+            (U128::zero(), U128::zero())
+        } else if self < b {
+            (U128::zero(), self)
         } else {
-            //TODO check if this can overflow?
-            let (q,r) = self.unconstrained_div(b * U128::from_u64s_le(2, 0));
+            let (q,r) = if b.hi as u64 >= pow63 as u64 {
+                // The result of multiplication by 2 would overflow
```

Comment:
I don't think this is correct, you should consider the case when self == b

Reply:
Nice catch, thanks